### DISCUSSION TITLE
Rule to blink LED?

### DISCUSSION THREAD

**Discussion by inventhouse** [arendst/Tasmota#10371] [Q&A] [answered]
Is there a good rule recipe for making the LED on the device blink repeatedly?

I'm trying to compose a ruleset for an auto-off countdown timer; I started with the [cookbook version](https://tasmota.github.io/docs/Rules/#time-delayed-auto-off-switch), but I'd like it to start blinking the LED when it's getting close to shutting off.

As a simple step in that direction, I added a rule I _think_ should blink the LED while the timer is counting down, but it's not working:

```
Rule1
  ON button1#state DO Backlog Power1 %value%; RuleTimer1 %mem1% ENDON
  ON Rules#Timer=1 DO Power1 off ENDON
  ON ruletimer#T1|2 DO LedPower1 2 ENDON
```

When the rule is running, the `ruletimer` command does, indeed, show the counting timer as `T1`:

> 17:25:30 CMD: ruletimer
> 17:25:30 RSL: stat/tasmota/RESULT = {"T1":13,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
 
But the rule based on that command doesn't ever seem to trigger; I am able to toggle the LED myself with the `LedPower1 2` command at the console.

Thanks for your help!

PS: I am running [Tasmota 9.2.0](http://ota.tasmota.com/tasmota/release-9.2.0/tasmota.bin.gz), the device is a basic [Aoycocr X5P](https://smile.amazon.com/Smart-Plug-Aoycocr-Control-Function/dp/B07N1JPPXK/ref=sr_1_2?dchild=1&keywords=aoycocr%2Bx5%2Bsmart%2Bplug&qid=1609606120&sr=8-2&th=1) smart plug

**Comment by barbudor** [comment, marked as answer]
Hi,
the trigger for ruletimer is the one you used in the 2nd rule : `on rules#timer=1`
There is no possibility to trigger with the modulo because the rule engine is called only when the timer reach 0, not every second.
You should use a 2nd ruletimer for the blinking :
```
Rule1 
  ON button1#state DO Backlog Power1 %value%; RuleTimer1 %mem1%; RuleTimer2 2 ENDON
  ON Rules#Timer=2 DO Backlog RuleTimer2 2; LedPower1 2 ENDON
  ON Rules#Timer=1 DO Backlog RuleTimer2 0; Power1 off ENDON
```
Should probably do the trick

**Reply by inventhouse** [reply]
Thanks!  The [docs say](https://tasmota.github.io/docs/Rules/#rule-trigger) "Every command with a JSON payload response has an associated rule trigger", so it seems to me like `ON ruletimer#T1|2 DO LedPower1 2 ENDON` should call `ruletimer`, extract `T1`, and check the modulo; I still find the commands and rule system rather confusing.

At any rate, I worked out a trick to do it:
```
ON Var1#State>0 DO Backlog LedPower1 2; Mult1 1 ENDON
```
When `Var1` is set to `1`, the LED starts blinking; when set to `0` it stops.  It seems if you want to use `ON Var1#State=1` you need to set `CalcRes 0` or otherwise handle the `Mult1` making Var1 no longer an int.

**Reply by barbudor** [reply]
You wrote:
_"The docs say "Every command with a JSON payload response has an associated rule trigger", "_

This is correct. 
The command Ruletimer issue a JSON such as `{"T1":13,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}`
And it can be captured by a rule as `on T1 ...`

_"so it seems to me like ON ruletimer#T1|2 DO LedPower1 2 ENDON should call ruletimer, extract T1, and check the modulo; I still find the commands and rule system rather confusing."_
*
This part is incorrect. 
The ON part of a rule do not issue a command. It reacts/capture/handle a response to a command.

So the `{ T1:...}`JSON is issued only when the RuleTimer command is entered.

With this rule
`ON Var1#State>0 DO Backlog LedPower1 2; Mult1 1 ENDON`

It looks like you have created a rule-loop that triggers itself.
I'm afraid that you are creating a monster and your Tasmota may becomes unresponsive
```
...
22:54:04.589 RUL: VAR1#STATE>0 performs "Mult1 1"
22:54:04.596 MQT: stat/dev-4119/MULT = {"Mult1":"1"}
22:54:04.641 RUL: VAR1#STATE>0 performs "Mult1 1"
22:54:04.649 MQT: stat/dev-4119/MULT = {"Mult1":"1"}
22:54:04.692 RUL: VAR1#STATE>0 performs "Mult1 1"
22:54:04.699 MQT: stat/dev-4119/MULT = {"Mult1":"1"}
22:54:04.745 RUL: VAR1#STATE>0 performs "Mult1 1"
22:54:04.753 MQT: stat/dev-4119/MULT = {"Mult1":"1"}
22:54:04.795 RUL: VAR1#STATE>0 performs "Mult1 1"
22:54:04.803 MQT: stat/dev-4119/MULT = {"Mult1":"1"}
...
...
```

**Reply by inventhouse** [reply]
Ah, thanks for the clarification!  It was where [the docs](https://tasmota.github.io/docs/Rules/#rule-trigger) described this as `<command>#level1#level2#levelN` that confused me, perhaps instead of "command" it should say "top-level" or somesuch; should I file an issue?

That's what it's doing; so far I've tried letting it blink for a couple minutes at a time and haven't had any trouble, but thanks for the warning; I'll add a note to [my write-up](https://github.com/arendst/Tasmota/discussions/10380) and mark your answer above.

**Reply by 19cool53** [reply]
a less complex solution to make the led blink:
-----------------------------------------------------------------
rule1
 on system#boot do backlog
  var1 1;
  pulsetime1 1;
  pulsetime2 10;
 endon
 on power2#state=0 do backlog
  power2 %var1%;
  power1 1
 endon
 on Time#minute do backlog
  power2 1
 endon
 
 pulsetime1: 0.1 sec on time on power1 ( == led)
 pulsetime2  1 sec on-time on (virtual) power2
 if power2 switches off:
 -- power1 is switched on, off after 0.1 sec
 -- power2 re-starts it self
if  var1=0
-- power2 does not start, led will not blink 
 time=minute
 -- is a help to start if var1 was changed
 -- or if something went wrong
-- or to blink every minute while var1==0,  (use power2 1)
-- otherwise power2 %var1%   (no blinks at all)

**Reply by sfromis** [reply]
Simplest way to make a led blink is to config it as a Relay, and using commands `Power blink` along with `BlinkTime` and `BlinkCount` for config. No rules.
